catlearn/utils/metrics.py

```diff
@@ -0,0 +1,190 @@
+import os
+from typing import Any, Tuple, Optional, FrozenSet, List, Mapping
+from collections import defaultdict
+from itertools import chain, zip_longest
+import math
+
+import wandb
+import torch
+from torch.nn.functional import pad
+
+from .numerics import Tsor, get_index_in_list
+from ..composition_graph import (
+    NodeType, ArrowType, DirectedGraph, CompositeArrow,
+)
+from ..relation_cache import RelationCache
+from ..categorical_model import TrainableDecisionCatModel
+
+
+ArrowHypothesisTriplet = Tuple[NodeType, NodeType, FrozenSet[ArrowType]]
+PartialArrowHypothesisTriplet = Tuple[
+        Optional[NodeType], Optional[NodeType],
+        Optional[FrozenSet[ArrowType]],
+]
+TripletMatchPair = Tuple[
+    ArrowHypothesisTriplet,
+    PartialArrowHypothesisTriplet,
+]
+
+
+def compute_training_metrics(
+    cache: RelationCache[NodeType, ArrowType],
+    matched: DirectedGraph[NodeType],
+) -> Mapping[str, Any]:
+    """
+        Log some metrics from a training step
+    """
+    # costs
+    total = sum(
+        sum(elem for _, elem in costs.values())
+        for costs in matched.edges.values())
+    nb_labels = sum(len(costs) for costs in matched.edges.values())
+    cost_per_label = total/max(nb_labels, 1) + cache.causality_cost
+
+    # arrow by order and causality check
+    used_arrows = set(chain(*((
+        arr.suspend(src, tar) for (arr, _) in content.values())
+        for src, tar, content in matched.edges(data=True))))
+    causal_arrows = set(cache.arrows(include_non_causal=False))
+    noncausal_arrows = set(cache.arrows()) - causal_arrows
+
+    arrows = defaultdict(lambda: {'used': 0, 'causal': 0, 'noncausal': 0})
+    for arr in used_arrows:
+        arrows[len(arr)]['used'] += 1
+    for arr in causal_arrows:
+        arrows[len(arr)]['causal'] += 1
+    for arr in noncausal_arrows:
+        arrows[len(arr)]['noncausal'] += 1
+
+    return {
+        'nb_labels': nb_labels,
+        'total_match_cost': total,
+        'total_causality_cost': cache.causality_cost,
+        'cost_per_label': cost_per_label,
+        'arrow_numbers': arrows,
+}
+
+from tqdm import tqdm
+
+def compute_eval_triplet_ranks(
+    cache: RelationCache[NodeType, ArrowType],
+    triplets: List[TripletMatchPair],
+    default_rank: int,
+    max_rank: Optional[int] = None,
+) -> Mapping[ArrowHypothesisTriplet, int]:
+    """
+    Compute the rank of the true triplet in the list of ranked hypothesis
+    generated from the partial triplet
+    - default_rank: The rank used if the triplet is not found
+    - max_rank: only ranks up to max_rank are computed.
+    Hence will default to default_rank above it
+    (if None all ranks are taken into account).
+    """
+    # compute ranks for each triplet
+    rank_lists = {
```

Comment:
You can fuse the two loop / iteration in one imho. That way we would need less memory since wa won't have to store all rank_lists